xcube_clms/preload.py

```diff
@@ -0,0 +1,530 @@
+import os
+import tempfile
+import threading
+import time
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor
+from typing import Any
+
+import fsspec
+import xarray as xr
+from rioxarray import rioxarray
+from tqdm.notebook import tqdm
+from xcube.core.store import new_data_store
+
+from xcube_clms.api_token import CLMSAPIToken
+from xcube_clms.constants import (
+    JSON_TYPE,
+    LOG,
+    UNDEFINED,
+    COMPLETE,
+    TITLE_KEY,
+    NOT_SUPPORTED_LIST,
+    FULL_SOURCE_KEY,
+    ITEMS_KEY,
+    DATASET_DOWNLOAD_INFORMATION_KEY,
+    SOURCE_KEY,
+    PATH_KEY,
+    UID_KEY,
+    DOWNLOAD_ENDPOINT,
+    ID_KEY,
+    ACCEPT_HEADER,
+    CONTENT_TYPE_HEADER,
+    TASK_STATUS_ENDPOINT,
+    STATUS_KEY,
+    DATASETS_KEY,
+    DATASET_ID_KEY,
+    STATUS_PENDING,
+    PENDING,
+    STATUS_COMPLETE,
+    DOWNLOAD_URL_KEY,
+    RESULTS,
+    FILENAME_KEY,
+    NAME_KEY,
+    DOWNLOAD_AVAILABLE_TIME_KEY,
+    ORIGINAL_FILENAME_KEY,
+    FILE_ID_KEY,
+    RETRY_TIMEOUT,
+    STATUS_CANCELLED,
+    GEO_FILE_EXTS,
+    DATA_ID_SEPARATOR,
+    TASK_IDS_KEY,
+    TASK_ID_KEY,
+    CANCELLED,
+)
+from xcube_clms.utils import (
+    make_api_request,
+    get_response_of_type,
+    get_authorization_header,
+    get_dataset_download_info,
+    build_api_url,
+    has_expired,
+    spinner,
+    find_easting_northing,
+    cleanup_dir,
+)
+
+
+class PreloadData:
```

Comment:
remember Normans issue: https://github.com/xcube-dev/xcube/issues/1093

In the same sense we need to come up with a general set of methods, which defines the PreloadData class. 
So can you sort this out a little bit. This will be helpful when we want to abstract this to an abstract class for xcube.core.store. I suggest, in the most general sense this class should have the following methods: 

- `download_data`: which does the data download. Issuing a request in CLMS is specific to xcube-clms and should go into local methods starting with `_*`. The download data method can take a while, and thus shall support some progress viewing, which you already have. 
- `process_data`: this opens the data, and does the stacking in your case. 
- `monitor_cache`: this will give the user an overview, which datasets are prepared.

Are there more methods which are essential for this class?

Reply:
Yeah, I wanted to do this and have done this now.
I think for now, these methods are sufficient. Later, we will need to handle the TaskHandlers that allow the user to view the status of each request and the user can interact with it like cancelling the request.